src/pudl/transform/eia930.py

```diff
@@ -0,0 +1,134 @@
+"""Module to perform data cleaning functions on EIA930 data tables."""
+
+import pandas as pd
+from dagster import AssetOut, Output, asset, multi_asset
+
+import pudl
+from pudl.metadata.enums import EIA930_GENERATION_ENERGY_SOURCES
+
+logger = pudl.logging_helpers.get_logger(__name__)
+
+
+@multi_asset(
+    outs={
+        "core_eia930__hourly_balancing_authority_net_generation_by_energy_source": AssetOut(
+            io_manager_key="parquet_io_manager"
+        ),
+        "core_eia930__hourly_balancing_authority_operations": AssetOut(
+            io_manager_key="parquet_io_manager"
+        ),
+    },
+    compute_kind="pandas",
+)
+def core_eia930__hourly_balancing_authority_assets(
+    raw_eia930__balance: pd.DataFrame,
+):
+    """Separate raw_eia930__balance into net generation and demand tables.
+
+    Energy source starts out in the column names, but is stacked into a categorical
+    column. For structural purposes "interchange" is also treated as an "energy source"
+    and stacked into the same column. For the moment "total" (sum of all energy sources)
+    is also included, because the reported and calculated totals acrss all energy
+    sources have significant differences which should be further explored.
+    """
+    nondata_cols = [
+        "datetime_utc",
+        "balancing_authority_code_eia",
+    ]
+    # Select all columns that aren't energy source specific
+    operations = raw_eia930__balance[
+        nondata_cols
+        + list(
+            raw_eia930__balance.filter(
+                regex=r"(demand|interchange|net_generation_total)"
+            )
+        )
+    ].rename(columns=lambda x: x.replace("net_generation_total_", "net_generation_"))
+    # Select only the columns that pertain to individual energy sources. Note that for
+    # the "unknown" energy source there are only "reported" values.
+    netgen_by_source = (
+        raw_eia930__balance[
+            nondata_cols
+            + [
+                f"net_generation_{fuel}_{status}_mwh"
+                for fuel in EIA930_GENERATION_ENERGY_SOURCES
+                for status in ["reported", "adjusted", "imputed"]
+                if fuel != "unknown"
+            ]
+            + ["net_generation_unknown_reported_mwh"]
+        ]
+        .rename(
+            # Rename columns so that they contain only the energy source and the level
+            # of processing with the pattern: energysource_levelofprocessing so the
+            # column name can be split on "_" to build a MultiIndex before stacking.
+            lambda col: col.removeprefix("net_generation_").removesuffix("_mwh"),
+            axis="columns",
+        )
```

Comment:
okay this means that the `EIA930_GENERATION_ENERGY_SOURCES` cannot have any `_` in those strings. That seems like an important implied detail. Can you add a little check for that somewhere that yells if this is ever not the case?

Reply:
I've added an assertion where the global constant is defined, but I'm not sure that's what you had in mind here.

Reply:
that is fine! not blocking at all but adding a comment in this function about this implied assumption would be good.